engines/nancy/action/primaryvideo.cpp

```diff
@@ -0,0 +1,382 @@
+    uint16 numFlags = stream.readUint16LE();
+    
+    for (uint i = 0; i < numFlags; ++i) {
+        conditionFlags.push_back(ConditionFlag());
+        conditionFlags.back().read(stream);
+    }
+}
+
+bool PlayPrimaryVideoChan0::ConditionFlags::isSatisfied() const {
+    bool orFlag = false;
+
+    for (uint i = 0; i < conditionFlags.size(); ++i) {
+        const ConditionFlag &cur = conditionFlags[i];
+        
+        if (!cur.isSatisfied()) {
+            if (orFlag) {
+                return false;
+            } else {
+                orFlag = true;
+            }
+        }
+    }
+
+    if (orFlag) {
+        return false;
+    } else {
+        return true;
+    }
+}
+
+PlayPrimaryVideoChan0::~PlayPrimaryVideoChan0() {
+    _decoder.close();
+
+	if (activePrimaryVideo == this) {
+		activePrimaryVideo = nullptr;
+	}
+    
+    NancySceneState.setShouldClearTextbox(true);
+    NancySceneState.getTextbox().setVisible(false);
+}
+
+void PlayPrimaryVideoChan0::init() {
+    _decoder.loadFile(videoName + ".avf");
+    _drawSurface.create(src.width(), src.height(), _decoder.getPixelFormat());
+
+    RenderObject::init();
+    
+    NancySceneState.setShouldClearTextbox(false);
+}
+
+void PlayPrimaryVideoChan0::updateGraphics() {
+    if (!_decoder.isVideoLoaded()) {
+        return;
+    }
+
+    if (!_decoder.isPlaying()) {
+        _decoder.start();
+    }
+
+    if (_decoder.needsUpdate()) {
+        _drawSurface.blitFrom(*_decoder.decodeNextFrame(), src, Common::Point());
+        _needsRedraw = true;
+    }
+
+    RenderObject::updateGraphics();
+}
+
+void PlayPrimaryVideoChan0::onPause(bool pause) {
+    _decoder.pauseVideo(pause);
+
+    if (pause) {
+        registerGraphics();
+    }
+}
+
+void PlayPrimaryVideoChan0::readData(Common::SeekableReadStream &stream) {
+    uint16 beginOffset = stream.pos();
+
+    char name[10];
+    stream.read(name, 10);
+    videoName = Common::String(name);
+
+    stream.skip(0x13);
+
+    readRect(stream, src);
+    readRect(stream, _screenPosition);
+
+    char *rawText = new char[1500]();
+    stream.read(rawText, 1500);
+    UI::Textbox::assembleTextLine(rawText, text, 1500);
+    delete[] rawText;
+
+    sound.read(stream, SoundDescription::kNormal);
+    responseGenericSound.read(stream, SoundDescription::kNormal);
+    stream.skip(1);
+    conditionalResponseCharacterID = stream.readByte();
+    goodbyeResponseCharacterID = stream.readByte();
+    isDialogueExitScene = (NancyFlag)stream.readByte();
+    doNotPop = (NancyFlag)stream.readByte();
+    sceneChange.readData(stream);
+
+    stream.seek(beginOffset + 0x69C);
+
+    uint16 numResponses = stream.readUint16LE();
+    if (numResponses > 0) {
+        for (uint i = 0; i < numResponses; ++i) {
+            responses.push_back(ResponseStruct());
+            ResponseStruct &response = responses[i];
+            response.conditionFlags.read(stream);
+            rawText = new char[400];
```

Comment:
This buffer should get reused and allocated only once.

Reply:
Done.